amd/src/courseformat/tabtree.js

```diff
@@ -0,0 +1,211 @@
+            addAnchor.setAttribute("href", addLink);
+        }
+
+        // Do things that make the first row tabs match firstsectionlist.
+        const toptabslist = element.firstsectionlist ?? [];
+        const childtabslist = element.secondsectionlist ?? [];
+        let toptabs = this.element.querySelector('ul:first-of-type');
+        this._fixOrder(toptabs, toptabslist, this.tabs, childtabslist[this.activetab[0]].length > 1);
+
+        // And the second row tabs match secondsectionlist.
+        let childtabs = this.element.querySelector('ul:nth-of-type(2)');
+        if (childtabs) {
+            this._fixOrder(childtabs, childtabslist[this.activetab[0]], this.childtabs, false);
+        }
+    }
+
+    /**
+     * Fix/reorder the section or cms order.
+     *
+     * @param {Element} container the HTML element to reorder.
+     * @param {Array} neworder an array with the ids order
+     * @param {Array} allitems the list of html elements that can be placed in the container
+     * @param {boolean} hassubtree
+     */
+    _fixOrder(container, neworder, allitems, hassubtree) {
+
+        // Empty lists should not be visible.
+        if (!neworder.length) {
+            container.classList.add('hidden');
+            container.innerHTML = '';
+            return;
+        }
+
+        // Grant the list is visible (in case it was empty).
+        container.classList.remove('hidden');
+
+        // Move the elements in order at the beginning of the list.
+        neworder.forEach((itemid, index) => {
+            const section = this.reactive.stateManager.state.section.get(itemid);
+            const visible = (section.visible && section.available || section.section == 0)
+                && (neworder.length > 1 || hassubtree);
+            if (allitems[itemid] === undefined) {
+                // If we don't have an item, create it from the course index.
+                let selecta = "[data-id='" + itemid + "']";
+                let ciElement = document.querySelector(selecta);
+                let ciLink = ciElement.querySelector(" a.courseindex-link");
+                let data = {
+                    "active": 0,
+                    "inactive": 0,
+                    "link": [{
+                    "link": ciLink.getAttribute("href")
+                }],
+                    "title": ciLink.innerHTML,
+                    "text": '<div class="tab_content' + (visible ? '' : ' dimmed')
+                        + '" data-itemid="' + section.id + '">' + section.title + '</div>'
+                };
+                let tab = document.createElement("li");
+                allitems[itemid] = tab;
+                container.insertBefore(tab, container.lastElementChild);
+                Templates.render("format_multitopic/courseformat/tab", data).done(function(html) {
+                    Templates.replaceNode(item, html);
```

Comment:
It doesn't look like storing the tab in allitems worked.  (If you drag a second-level tab into the active top-level tab, then out, then back in again, it doesn't show properly.)  I think it needs to be stored here, after the replacement is made.  But it looks like this function doesn't return if it doesn't have a third argument.  Also, technically I think the variable item probably shouldn't be used here since it's declared further down (although this doesn't seem to be an issue in practice).
```suggestion
                    allitems[itemid] = Templates.replaceNode(tab, html, "")[0];
```